tests/MongoDB.Driver.Tests/Specifications/crud/CrudTestRunner.cs

```diff
@@ -19,55 +19,228 @@
+                var expectedEvents = test["expectations"].AsBsonArray.Cast<BsonDocument>().GetEnumerator();
+
+                while (actualEvents.Any())
+                {
+                    var actualEvent = actualEvents.Next();
+
+                    if (!expectedEvents.MoveNext())
+                    {
+                        throw new Exception($"Unexpected event of type: {actualEvent.GetType().Name}.");
+                    }
+                    var expectedEvent = expectedEvents.Current;
+                    AssertEvent(actualEvent, expectedEvent);
+                }
 
-            database.DropCollection(collection.CollectionNamespace.CollectionName);
-            collection.InsertMany(definition["data"].AsBsonArray.Cast<BsonDocument>());
+                if (expectedEvents.MoveNext())
+                {
+                    var expectedEvent = expectedEvents.Current;
+                    throw new Exception($"Missing event: {expectedEvent}.");
+                }
+            }
+        }
 
-            ExecuteOperation(database, collection, (BsonDocument)test["operation"], (BsonDocument)test["outcome"], async);
+        private DisposableMongoClient CreateDisposableClient(EventCapturer eventCapturer)
+        {
+            return DriverTestConfiguration.CreateDisposableClient(
+                (MongoClientSettings settings) =>
+                {
+                    settings.ClusterConfigurator = c =>
+                    {
+                        c = CoreTestConfiguration.ConfigureCluster(c);
+                        c.Subscribe(eventCapturer);
+                        c.ConfigureServer(ss => ss.With(heartbeatInterval: Timeout.InfiniteTimeSpan));
+                    };
+                });
         }
 
-        private void ExecuteOperation(IMongoDatabase database, IMongoCollection<BsonDocument> collection, BsonDocument operation, BsonDocument outcome, bool async)
+        private void DropCollection(string databaseName, string collectionName)
+        {
+            DriverTestConfiguration.Client.GetDatabase(databaseName).DropCollection(collectionName);
+        }
+
+        private void ExecuteOperation(IMongoClient client, IMongoDatabase database, IMongoCollection<BsonDocument> collection, BsonDocument operation, BsonDocument outcome, bool async)
         {
             var name = (string)operation["name"];
             var test = CrudOperationTestFactory.CreateTest(name);
+            bool isErrorExpected = operation.GetValue("error", false).AsBoolean;
 
             var arguments = (BsonDocument)operation.GetValue("arguments", new BsonDocument());
-            string reason;
-            if (!test.CanExecute(DriverTestConfiguration.Client.Cluster.Description, arguments, out reason))
+            test.SkipIfNotSupported(arguments);
+
+            test.Execute(client.Cluster.Description, database, collection, arguments, outcome, isErrorExpected, async);
+
+            if ((test.ActualError != null) != isErrorExpected)
```

Comment:
Compare an error state from the test and our error expectations from the test.
if there is a mismatch, throw the exception.